larq_compute_engine/tflite/kernels/quantization.cc

```diff
@@ -80,6 +85,27 @@ TfLiteStatus QuantizeEval(TfLiteContext* context, TfLiteNode* node) {
   } else if (input->type == kTfLiteInt8) {
     bitpack_tensor(GetTensorShape(input), GetTensorData<std::int8_t>(input),
                    input->params.zero_point, GetTensorData<TBitpacked>(output));
+  } else if (input->type == kTfLiteBool) {
+    // The strategy here is to interpret the input data as an unsigned integer
+    // (of the same width as the bool type for the target). We then call
+    // bitpacking, with a 'zero point' of 1. This means that the value with all
+    // zero bits will be bitpacked as bit 1, and all other values will be
+    // bitpacked as bit 0. Assuming that `false` is represented by a value with
+    // all zero bits, this gives the correct result of bitpacking `false` as bit
+    // 1 and `true` as bit 0.
```

Comment:
How do you feel about changing the MLIR pattern to make `true` correspond to float `-1` and `false` to float `+1`, i.e. the boolean is the value of the signbit, `true` when the float value is negative.

Then this kernel code might be simpler, in the sense that the boolean values correspond to the output bits directly.
See also the comment in `prepare_patterns_common.td`.

Reply:
I think that's slightly a question for the Larq API. We could for sure do that, and then require people to write quantisers as e.g. `tf.where(x > 5, minus_ones, ones)` but I feel like instinctively people having the true bits correspond to 1 feels more natural. The fact that we bitpack 1 as 0 and -1 as 1 to me feels like an implementation detail that we shouldn't expose like that, but that's not a belief I hold super strongly.

I'm also not 100% sure what the bit representation of `true` is — I can't find any standard that specifies it. I assume it's the byte `0x01`, but I guess could also be the byte `0xFF`? And if we do know what the bit representation is, can we assume it's the same for all targets? To be fair, at this point in the code I am assuming that `false` is represented by just zero bits, which also isn't specified anywhere, so perhaps the further assumption isn't an issue.

Reply:
I agree that its an implementation detail that shouldn't be exposed in that sense, and I agree that most people will find `true = +1, false = -1` more intuitive. Both forms should still be accepted by the converter (as discussed in the other comment).

I think that indeed the representation of `true` isn't standardized, it can be `1` or `-1=0xffffffff` etc, but `false` is probably always `0` even though that is also not specified.

Reply:
Ah okay I think I misunderstood your original comment. I'm completely on board with accepting both forms, and pushing through a TFL `logical_not` op if necessary :)

Then the question is which ought to be the 'canonical' form — since we can push through a `logical_not`, this boils down to whether we want to map `true` -> `0 bit` (which is what I do now) or map `true` -> `1 bit`, and which is easier in terms of kernel implementation.

I *think* that in any case we have to collect sign bits from an unsigned datatype, because as you say we don't know for sure if `true` is represented by `0x01` or `0xFF`. In the former case we have `signed_repr(false) < signed_repr(true)`, but in the latter case the opposite. This means that we would get different results from any operation that relies on collecting sign bits. However, as long as `false` is always represented by `0x00`, `unsigned_repr(false) < unsigned_repr(true)` always holds, and so collecting sign bits (with a zero-point of 1) is guaranteed to work and produce the correct result, no matter what the bit representation of `true` is. The natural format, using this approach, is to map `true` -> `0 bit`, in fact I don't see an easy way to do it for the other way round.

Reply:
You're right, this is the only method that's independent of the `bool` representation :+1:. Lets keep using this form as canonical form then.